### DISCUSSION TITLE
NSWindowCollectionBehaviorCanJoinAllSpaces with Ebitengine

### DISCUSSION THREAD

**Discussion by crgimenes** [hajimehoshi/ebiten#2948] [Q&A]
I created a program that aims to place a transparent, floating, and borderless window covering the entire screen. Then, I can send commands via Unix Domain Socket and draw on the screen (imitating the behavior of drawing on the screen with BASIC as we did in the 8-bit era.). It would be effortless to do this using Ebitengine, but I ran into a problem; I want the program to be visible in all "spaces", and that's the only thing I couldn't do in Ebitengine.

So, I wrote a small snippet of code in Objective-C to test the concept, which worked perfectly.
The only property I need is `NSWindowCollectionBehavior` with the value `NSWindowCollectionBehaviorCanJoinAllSpaces`, which makes the window visible in all spaces.

Here is a link to this property in my prototype:

https://github.com/crgimenes/NeoFrame/blob/master/screen/screen_macos.go#L31

If anyone has any idea how to do this in Ebitengine, it will help me a lot; writing the program in Go will be much simpler than in Objective-C.

**Comment by TotallyGamerJet** [comment]
In order to call the macOS libraries without OBJC you’ll need to use purego. However, I don’t think ebitengine exposes the native macOS window to be able to change that property.

**Reply by crgimenes** [reply]
I believe I need to interact with GLFW, which as far as I know is who creates the window. I haven't found a painless way to do this yet.

**Reply by TotallyGamerJet** [reply]
Yes in ebitengine GLFW is used to create the window. GLFW has `glfwGetCocoaWindow`. However, ebitengine doesn't expose this function so you can't call it.

**Reply by crgimenes** [reply]
It may be possible to capture the main window and then change the property.

```objc
[NSApplication sharedApplication];
[[NSRunLoop currentRunLoop] runMode:NSDefaultRunLoopMode beforeDate:[NSDate distantPast]];
NSWindow *window = [NSApplication sharedApplication].mainWindow;
[window setCollectionBehavior:NSWindowCollectionBehaviorCanJoinAllSpaces];
```

I'll try this approach and see if it works.

**Comment by crgimenes** [comment]
I tried capturing the window using Objective-C and changing the property with the function below and several variations of it; I can get the window, but changing the property does not have any effect.

```objective-c
#cgo CFLAGS: -x objective-c
#cgo LDFLAGS: -framework Cocoa

#import <Cocoa/Cocoa.h>

void setWindowCollectionBehavior() {
	dispatch_async(dispatch_get_main_queue(), ^{
		NSApplication *app = [NSApplication sharedApplication];
		NSWindow *window = app.mainWindow;
		[window setCollectionBehavior:NSWindowCollectionBehaviorCanJoinAllSpaces];
	});
}
```

Does anyone have any different suggestions?
